src/marqo/vespa/vespa_client.py

```diff
@@ -675,46 +675,77 @@ async def _update_document_async(self, semaphore: asyncio.Semaphore, async_clien
+            except JSONDecodeError as e:
+                if resp.status_code == 200:
+                    # A 200 response shouldn't reach here
+                    raise VespaError(f"Marqo received an unexpected response from the vector store: {resp.text} with 200 "
+                                     f"status_code. The response is not in the JSON format and Marqo can not decode "
+                                     f"it") from e
+                else:
+                    error_message = (f"Marqo has encountered an error while update this document to the vector store "
+                                     f"and the response can not be decoded. The document is not updated. The response "
+                                     f"text is: {resp.text}")
+                    return UpdateDocumentResponse(status=500, message=error_message, pathId=error_doc_path_id, id=doc_id)
 
     async def _feed_document_async(self, semaphore: asyncio.Semaphore, async_client: httpx.AsyncClient,
                                    document: VespaDocument, schema: str,
                                    timeout: int) -> FeedBatchDocumentResponse:
+        """An async method to feed a document to Vespa.
+
+        Note: This method is used by the async feed batch method to feed documents concurrently. Unhandled exceptions
+        will be raised in the main thread and leads a 500 error for the whole batch. Therefore, exceptions should be
+        handled gracefully in this method for the specific document. We should keep the error message as similar as the
+        Vespa error messages since this is a low level method. Overwrite the error message in higher level methods.
+
+        Exceptions that are handled in this method:
+        1. httpx.RequestError: We convert this error to a 500 error for the specific document and put 'Network Error' in
+        the message.
+        2.
+
+        Raises:
+            VespaError: If the Vespa response is 200 but the response can not be decoded.
+
+        Returns:
+            FeedDocumentResponse object
+        """
         doc_id = document.id
         data = {'fields': document.fields}
 
         async with semaphore:
             end_point = f'{self.document_url}/document/v1/{schema}/{schema}/docid/{doc_id}'
+            # Handle httpx.RequestError
             try:
                 resp = await async_client.post(end_point, json=data, timeout=timeout)
-            except httpx.HTTPError as e:
-                raise VespaError(e) from e
-
-        try:
-            # This will cover 200 and document-specific errors. Other unexpected errors will be raised.
-            return FeedBatchDocumentResponse(**resp.json(), status=resp.status_code)
-        except JSONDecodeError as e:
-            if resp.status_code == 200:
-                # A 200 response shouldn't reach here
-                raise VespaError(f'Unexpected response: {resp.text}') from e
+            except httpx.RequestError as e:
```

Comment:
In httpx, `HTTPError` is the base class for `RequestError` and `HTTPStatusError`. We changed to only capture `RequestError` here, `HTTPStatusError` will be raised to the outer layer. Is this intended?

Reply:
It seems that `HTTPStatusError` can only be raised by calling res.raise_for_status. And we have our own logic to raise for status and was handled below. So I think it should be OK.